module/state_synchronization/indexer/indexer_core.go

```diff
@@ -328,42 +340,24 @@ func (c *IndexerCore) indexRegisters(registers map[ledger.Path]*ledger.Payload,
 // IndexCollection handles the response of the collection request made earlier when a block was received.
 // No errors expected during normal operations.
 func IndexCollection(
+	lctx lockctx.Proof,
 	collection *flow.Collection,
 	collections storage.Collections,
-	transactions storage.Transactions,
 	logger zerolog.Logger,
 	collectionExecutedMetric module.CollectionExecutedMetric,
 ) error {
 
-	light := collection.Light()
-
-	collectionExecutedMetric.CollectionFinalized(light)
-	collectionExecutedMetric.CollectionExecuted(light)
-
 	// FIX: we can't index guarantees here, as we might have more than one block
 	// with the same collection as long as it is not finalized
 
 	// store the light collection (collection minus the transaction body - those are stored separately)
 	// and add transaction ids as index
-	err := collections.StoreLightAndIndexByTransaction(&light)
+	light, err := collections.StoreAndIndexByTransaction(lctx, collection)
 	if err != nil {
-		// ignore collection if already seen
-		if errors.Is(err, storage.ErrAlreadyExists) {
-			logger.Debug().
-				Hex("collection_id", logging.Entity(light)).
-				Msg("collection is already seen")
-			return nil
-		}
 		return err
 	}
 
-	// now store each of the transaction body
-	for _, tx := range collection.Transactions {
```

Comment:
`StoreAndIndexByTransaction` will store the transaction internally, so that txs and collection can be saved in the same batch update.